components/mediation/data-publishers/org.wso2.micro.integrator.observability/src/main/java/org/wso2/micro/integrator/observability/handler/metrichandler/MetricHandler.java

```diff
@@ -0,0 +1,388 @@
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.SynapseConstants;
+import org.apache.synapse.config.SynapsePropertiesLoader;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.rest.API;
+import org.apache.synapse.rest.RESTConstants;
+import org.apache.synapse.rest.RESTUtils;
+import org.apache.synapse.transport.nhttp.NhttpConstants;
+import org.wso2.config.mapper.ConfigParser;
+import org.wso2.micro.integrator.core.internal.MicroIntegratorBaseConstants;
+import org.wso2.micro.integrator.observability.handler.metrichandler.prometheus.reporter.PrometheusReporter;
+import org.wso2.micro.integrator.observability.handler.util.MetricConstants;
+
+import java.util.Map;
+
+/**
+ * Class for extracting metric information by wrapping the implementation and
+ * there by allows the implementation to plug into the handler in runtime.
+ */
+public class MetricHandler extends AbstractExtendedSynapseHandler {
+
+    private static Log log = LogFactory.getLog(MetricHandler.class);
+    private static final String METRIC_REPORTER = "metric_reporter";
+    private static final String DELIMITER = "/";
+    private static final String EMPTY = "";
+
+    private MetricReporter metricReporterInstance;
+    int serviceInvokePort;
+
+    private static final String SERVER_PORT_OFFSET = System.getProperty(MetricConstants.PORT_OFFSET);
+    private static final String HOST = System.getProperty(MicroIntegratorBaseConstants.LOCAL_IP_ADDRESS);
+    private static final String PORT = System.getProperty(MetricConstants.HTTP_PORT);
+    private static final String JAVA_VERSION = System.getProperty(MetricConstants.JAVA_VERSION);
+    private static final String JAVA_HOME = System.getProperty(MetricConstants.JAVA_HOME);
+    private int internalHttpApiPort = getInternalHttpApiPort();
+
+    @Override
+    public boolean handleServerInit() {
+        getMetricReporter();
+        metricReporterInstance.initMetrics();
+
+        metricReporterInstance.serverUp(HOST, PORT, JAVA_HOME, JAVA_VERSION);
+        return true;
+    }
+
+    /**
+     * Load the MetricReporter class from the deployment.toml file if a user has defined a MetricReporter.
+     * Use default PrometheusReporter if the user hasn't defined a MetricReporter or an error occurs
+     * during custom MetricReporter class invocation.
+     */
+    private MetricReporter getMetricReporter() {
+        Map<String, Object> configs = ConfigParser.getParsedConfigs();
+        Object metricReporterClass = configs.get(MetricConstants.METRIC_HANDLER + "." + METRIC_REPORTER);
+        Class loadedMetricClass;
+
+        if (metricReporterClass != null) {
+            try {
+                loadedMetricClass = Class.forName(metricReporterClass.toString());
+                metricReporterInstance = (MetricReporter) loadedMetricClass.newInstance();
+                log.debug("The class " + metricReporterClass + " loaded successfully");
```

Comment:
please check whether debug is enabled  , in other places as well where we do string concatenation. this will affect perf.

Reply:
Added the whether debug is enabled check